### PR TITLE
Add custom metrics and logging options

### PR DESCRIPTION
**Pull request by RemyLau**
Hi, I've been using GraphGym in PyG for a while now, mainly for large-scale evaluation on many datasets. Here are a few things that I think would be really helpful to have, which are the main changes in this PR.
1. **Option to disable model checkpoint**
2. **Option to disable evaluation during training epochs**
3. **Option for using customized metrics**

### CODE REVIEW

File: torch_geometric/graphgym/config.py [hunk 1]
```diff
@@ -229,9 +229,15 @@ def set_cfg(cfg):
     # Evaluate model on test data every eval period epochs
     cfg.train.eval_period = 10
 
+    # Option to skip training epoch evaluation
```
Comment (JiaxuanYou) [inline torch_geometric/graphgym/config.py:232]: I think adding a configuration that selects the customized statistics could be helpful. Of course, we can also assume that all the registered customized statistics will be used. Please let me know if what do you think.

File: test/graphgym/example_node.yml [hunk 2]
```diff
@@ -11,8 +11,10 @@ dataset:
   edge_encoder_name: Bond
 train:
   batch_size: 128
-  eval_period: 1
+  eval_period: 2
```
Comment (JiaxuanYou) [inline test/graphgym/example_node.yml:14]: If this is change is not needed, please keep the default configs unchanged, so that users can reproduce the examples in older versions of PyG GraphGym.

**Review by JiaxuanYou — COMMENTED** [review]
Thanks for your contribution! Please see my comments attached

File: torch_geometric/graphgym/train.py [hunk 3]
```diff
@@ -69,13 +70,14 @@ def train(loggers, loaders, model, optimizer, scheduler):
     split_names = ['val', 'test']
     for cur_epoch in range(start_epoch, cfg.optim.max_epoch):
         train_epoch(loggers[0], loaders[0], model, optimizer, scheduler)
-        loggers[0].write_epoch(cur_epoch)
+        if is_train_eval_epoch(cur_epoch):
```
[the remark this answers is not in the document]
Comment (RemyLau) [reply torch_geometric/graphgym/train.py:73]: Thanks for your reply! I've been using GraphGym to evaluate some large datasets (node-level tasks) including the OGBN ``arxiv`` and ``proteins``. Due to a large number of nodes in these graphs, the computation of auroc is very expensive and I've observed a severe GPU usage bottleneck (near 0% usage) because of this. Skipping the training epoch evaluation could effectively remedy this issue. Furthermore, I don't see the evaluations for the training epoch being used anywhere in the downstream analysis, so I thought it might not hurt to disable it in favor of improving the GPU usage.

It has also been suggested to look into the [torchmetric](https://torchmetrics.readthedocs.io/en/latest/) in https://github.com/pyg-team/pytorch_geometric/issues/3349#issuecomment-946354517, the metric refactoring in this PR should also enable this in a future PR.

File: test/graphgym/example_node.yml (hunk 2 above)
Comment (RemyLau) [reply test/graphgym/example_node.yml:14]: This is modified to see if the the ``cfg.train.skip_train_eval`` option is functioning correctly. If ``cfg.train.eval_period`` is set to 1, there's no difference regardless of whether ``skip_train_eval`` is ``True`` or ``False``.

I could also create a new copy of this yml and keep the ``eval_peirod`` in the original yml to be 1 for the purpose of reproducing older examples. Let me know if you prefer this option.

File: torch_geometric/graphgym/config.py (hunk 1 above)
Comment (RemyLau) [reply torch_geometric/graphgym/config.py:232]: Yeah, I think it's a good idea to specify which custom metric group to use.

In the [current setup](https://github.com/RemyLau/pytorch_geometric/blob/bed926723b5ef7a4e261522e5ba56e019c7e9917/torch_geometric/graphgym/logger.py#L163-L168), I tried to follow the GraphGym code design language for loading custom functions (the example I had in mind was the [``torch_geometric.graphgym.loadder.py``](https://github.com/RemyLau/pytorch_geometric/blob/bed926723b5ef7a4e261522e5ba56e019c7e9917/torch_geometric/graphgym/loader.py#L164-L168)):
1. Look through all relevant registered custom functions and return immediately if a custom function returns nontrivial results
2. If the above failed, use the default graphgym functions

Going back to the point of specifying which custom metric group to use, maybe one possible strategy could be:
1. Specify the desired custom metric group to use as ``cfg.custom_metric`` (set to ``'none'`` by default)
2. Instead of looking through all registered custom metric functions, find the one with the matched key of ``cfg.custom_metric``. And maybe print warning message if no matched key for ``cfg.custom_metric`` was found and ``cfg.custom_metric != 'none'`` and fall back to using the default implemented metrics.

Let me know what you think

File: torch_geometric/graphgym/train.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (JiaxuanYou) [reply torch_geometric/graphgym/train.py:73]: Thanks for explaining! This makes sense. And indeed, future PRs on `torchmetric` is welcomed!

File: test/graphgym/example_node.yml (hunk 2 above)
Comment (JiaxuanYou) [reply test/graphgym/example_node.yml:14]: I see. This makes sense. I think having two examples may not be necessary, so you can go ahead with this change

File: torch_geometric/graphgym/config.py (hunk 1 above)
Comment (JiaxuanYou) [reply torch_geometric/graphgym/config.py:232]: I think this API sounds great! You could further make `cfg.custom_metric` to be a list, since a user may want to use multiple customized metrics. If the list is empty, then no customized metric is used.

File: test/graphgym/test_graphgym.py [hunk 4]
```diff
@@ -75,3 +76,51 @@ def test_run_single_graphgym():
 
     shutil.rmtree(cfg.out_dir)
     shutil.rmtree(cfg.dataset.dir)
+
+
+def trivial_metric(true, pred, task_type):
+    return {'trivial': 1}
+
+
+def test_run_trivial_graphgym():
```
[the remark this answers is not in the document]
Comment (RemyLau) [reply test/graphgym/test_graphgym.py:85]: Hi @JiaxuanYou I looked a little bit more into the ``update_stats`` function and I don't think it will call the custom metrics directly. In my current setup, which is the same as that in the original setup, the ``task_stats`` (i.e. the metrics) are updated during the function call of ``write_epoch`` using ``self._train`` and ``self._pred`` as input, both of which are set during ``update_stats``.

So ``update_stats``does not directly call ``trivial`` but sets the ``self._train`` and ``self._pred`` instead. Let me know if I'm understanding this wrong or if you want me to handle the custom metric function calls differently.

**Comment by rusty1s** [comment]
@RemyLau @JiaxuanYou Any status updates on this PR? Is this blocked by anything or ready to get merged?

**Comment by RemyLau** [comment]
Hi @rusty1s, thanks for bringing this up again! The most recent commit should resolve a major comment by @JiaxuanYou about selecting a group of custom metrics to be used, so it should be ready to be reviewed.

Because it's been a long time since the last commit, there are multiple conflicts with the main branch. Should I do a rebase and force push to update this PR?

**Comment by rusty1s** [comment]
Yes, that would be great :)

**Comment by RemyLau** [comment]
I've just rebased the branch back to master so all conflicts are fixed now

**Comment by rusty1s** [comment]
Thank you @RemyLau.

@JiaxuanYou Can you finish your review? I'm happy to merge it afterwards.

**Comment by JiaxuanYou** [comment]
Sorry for missing the thread. Yes, the new functionality of enabling a list of customized metrics looks good to me!

**Review by JiaxuanYou — APPROVED** [review]

**Comment by RemyLau** [comment]
Great! Thanks a lot!

File: test/graphgym/test_graphgym.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rusty1s) [reply test/graphgym/test_graphgym.py:74]: I think what @JiaxuanYou means is that we never actually test if `trivial_metric` is executed. To fix, we might want to add a counter to `trivial_metric`, similar to [here](https://github.com/pyg-team/pytorch_geometric/blob/master/test/nn/conv/test_message_passing.py#L195-L218).

**Comment by RemyLau** [comment]
Hi @rusty1s, I've attempted to make the changes you suggested: 1) test the execution of ``trivial_metric`` using a counter, 2) combine the two test functions. Let me know if this looks good to you

**Review by rusty1s — APPROVED** [review]
Thank you! I modified the tests a little further to make use of `pytest.mark.parametrize`.
